### DISCUSSION TITLE
Documentation improvements

### DISCUSSION THREAD

**Discussion by orefalo** [lmorg/murex#720] [Ideas]
Sharing my upcoming TODO

UserGuide
- [ ] Add a page section about data types
- [ ] Complete section about pipes, add redirection (<!null> first parameter), `<env:DOCGEN_TARGET=vuepress>` https://dev.murex.rocks/commands/namedpipe.html#detail is a good start
- [x] Add Job control page
- [ ] Introduce expression vs evaluations
- [ ] Add a page that explains arrays and the difference between all the a* builtins
- [ ] Add a page about config, look for bits and pieces across the doc
    * config set shell color false
- [ ] rename 'Profile Files' 'StartUp Files' - not everyone knows what a Profile is. https://dev.murex.rocks/user-guide/profile.html
- [x] highlight cool modules in the doc
-  CANCELED: ensure fn() practices are properly captured in the doc (https://github.com/lmorg/murex/issues/714?notification_referrer_id=NT_kwDNy3WwNzU5NTYxOTk0Mzo1MjA4NQ)
- [x] fix typo, it's starship, not starfish
- [ ] provide details about the grammar used in `regexp` - https://dev.murex.rocks/commands/regexp.html
- [ ] `[ (range)` must be renamed `[..range]`
- [ ] `[ (index)` must be renamed `[index]`
- [ ] `( (brace quote)` must be renamed `(quote)`
- [ ] `@[` is a synonym for `[..range]` - create s dummy page or redirection
- [ ] Add operator pages for `==` `!=` `<=` `>`...etc

Builtins

- [ ] Move operators to the operators section ( ! ( [ [[ <> >> |> <stdin> ...
- [ ] Create operator pages for == != ? (which ?> should be better)
- [ ] Move deprecated stuff to a new section, 'Deprecated'
    * look for ref in the doc and aggregate
    * add a page about deprecated syntaxes, ie `builtin:`
- [ ] Rework function page, point to datatype page

Marketing
- [ ] Redo terminal screenshots/recordings on the front page
- [ ] Add `SQL Inlining` to the landing page

3rd Party
- [ ] Work with `brew` team to have them output Murex code - this should remove this ugly zsh dependency
- [ ] Work with `starship` to output murex code
- [ ] Work with `pnpm` to output murex code

Others
- [x] add start history to README.md https://api.star-history.com/svg?repos=lmorg/murex&type=Date

Optimization
- [ ] Find a way for ViewPress to order search results as Builtins, Operators, UserGuide, *

Clean up
- [x] remaining `function:` - https://dev.murex.rocks/commands/function.html#using-parameterized-variable-names

**Comment by orefalo** [comment]
To fix as well

```
2023/09/06 14:52:54 [WARNING] which: Cannot find document with the ID `commands/type`
2023/09/06 14:52:54 [WARNING] which: Cannot find document with the ID `commands/type`
2023/09/06 14:52:54 [WARNING] which: Cannot find document with the ID `commands/type`
2023/09/06 14:52:54 [WARNING] which: Cannot find document with the ID `commands/type`
2023/09/06 14:52:54 [WARNING] which: Cannot find document with the ID `commands/type`
2023/09/06 14:52:54 [WARNING] onprompt: Cannot find document with the ID `events/onkeypress`
2023/09/06 14:52:54 [WARNING] onprompt: Cannot find document with the ID `events/onkeypress`
2023/09/06 14:52:54 [WARNING] onprompt: Cannot find document with the ID `events/onkeypress`
```

**Reply by lmorg** [reply]
I've fixed the `commands/type` warning (basically just created a `type` builtin like I'd been planning to for a while)

**Comment by orefalo** [comment]
Could you help me on this item - "grammar used in `regexp`"
What tool, library are you using to implement the regexp? Is it golang regexp features?
If you have pointers to their docs, it would help close this item.
Thank you

**Reply by lmorg** [reply]
I'm just using the native Go regexp libraries because there's no value in reinventing the wheel here. That library is an implementation of RE2 which, unsurprisingly, is one designed by Russ Cox.

The syntax is documented at https://github.com/google/re2/wiki/Syntax
The white paper https://swtch.com/~rsc/regexp/regexp1.html

The syntax has a few quirks from what I've seen in, for example, Perl in that flags are defined at the start of the expression inside parenthesis rather than after the expression. Personally I don't much like that particular feature of the syntax but aside from that, the regex engine hasn't otherwise caused me any issues.

**Comment by orefalo** [comment]
I am trying to test locally, looking at the CICD I came up with fallow script

```
cd murex
go build -v github.com/lmorg/murex/utils/docgen
./docgen <env:DOCGEN_TARGET=vuepress>  -panic -warning -config gen/docgen.yaml
cp -v *.md *.svg docs/
cp -Rv gen/vuepress docs/.vuepress
pnpm i
pnpm run docs:build
./node_modules/.bin/vuepress dev
```

but the generated site lacks most of the content !#!$!@
any idea what I could be missing

**Reply by lmorg** [reply]
I'm working on a build where you can automatically merge into a branch called `website`, which will then publish to something like `https://website.murex.rocks`

For now, `./docgen -panic -warning -config gen/docgen.yaml` will generate the markdown docs. I'm not familiar enough with vuepress to say how it should be built there (i just used your code verbatim) but I suspect one piece you might be missing is an env var. `docgen` will compile slightly different markdown documents depending on what the target is (eg it for github or vuepress?).

If you're trying to create markdown docs for vuepress to ingest then you'll need the following env var set: `DOCGEN_TARGET=vuepress`. eg (in murex):
```
./docgen <env:DOCGEN_TARGET=vuepress> -panic -warning -config gen/docgen.yaml
```
(this will be why you see sections of files with content like `{{ if (env "DOCGEN_TARGET=vuepress" }}` or similar to hide markdown TOC from vuepress or add in vuepress specific extras.

**Reply by orefalo** [reply]
I assume `<env:DOCGEN_TARGET=vuepress>` is an undocumented feature of murex ? ;-/

**Reply by lmorg** [reply]
yeah....

so much so i haven't even come up for a good name for it yet. Possibly a "command modifier".

**Comment by orefalo** [comment]
I am starting the page about data types

What data types are defined in Murex?
 int, string, str, json, array

**Reply by orefalo** [reply]
Also what is the difference between `format` and `cast`

**Reply by tiymat** [reply]
These are the data types I know of. (I was surprised to find you can't query `runtime` for a list of all types, unless I'm missing something.)
`bool`, `columns`, `commonlog`, `csv`, `float`, `generic`, `int`, `json`, `jsonc`, `jsonl`, `null`, `num`, `path`, `paths`, `qs`, `str`, `string`, `toml`, `yaml`

My understanding of `cast` and `format` is `cast` is used to tell Murex that the underlying data is of a given type, whereas `format` is used to convert data to a different type.

So for example, here `cast` works because the actual content of `abc` is already JSON
```
murex » out '{ "foo": "bar", "bar": "baz"}' -> set abc
murex » runtime --variables -> ['abc'] -> ['DataType']
str
murex » out '{ "foo": "bar", "bar": "baz"}' -> cast json -> set abc
murex » runtime --variables -> ['abc'] -> ['DataType']
json
```

If we wanted the data in YAML, we would need `format`, because we're converting the data to a new data type.
```
murex » out '{ "foo": "bar", "bar": "baz"}' -> cast json -> format yaml
bar: baz
foo: bar
```

Hope that was helpful!

**Reply by lmorg** [reply]
tiymat is spot on.

It's worth noting that there is a difference between primitives and non-primitives. Primitives would be types that could be described as the atomic (for want a better term) parts of a data structure. This is better explained in code.

`num`, `str`, `bool`, etc are all data types. As are `json`, `toml`, `qs` etc. But if you wanted to dissect an object or array (whether that structure is `json` or any other data format), then you would get a `num`, `str`, `bool`, etc...ie a primitive.

At this point in time, there is no way to have an object or array value be anything other a primitive or the same data type for that structure (eg in a nested JSON object, a value could be an array and in that case that array's data type is also JSON).

Where this limitation is most visible is types like `paths`. `paths` is an array of `str` formatted like `$PATH`, eg `str:str:str:str`. Since it's primary concern is storing paths -- hence its name -- you'd think an index of that should be `path`, but that's not currently possible because `path` is not a primitive. Thus code like `$PATH.0.0.Exists` would fail whereas `$PATH.0 -> :path: [[/0/Exists]]` would return `true` because you're casting an `str` to `path`.

**Reply by lmorg** [reply]
The documentation around primitives needs to be rewritten though. It incorrectly states `json` is a primitive and doesn't explain what they are well (or even correctly). I'm not really sure how I got that doc so far wrong.

**Reply by orefalo** [reply]
@lmorg, it is as it is, and personally I think it's in excellent shape.
 You wrote a LOT, the doc only needs to be a bit better structured and updated (with the undocumented features).

Is this the page you are referring to in your last comments?
https://dev.murex.rocks/types/

I will most likely rewrite parts of it and move it to `user-guide`

**Reply by lmorg** [reply]
Yes that's the page. Technically that's an index page rather than a document, so the contents are in `docgen.yaml` (yeah, that design sucks. Docgen was thrown together quickly to make it easier for me to write docs and in fairness it did succeed in that. But the rushed nature it was thrown together certainly shows).

If you knock up a new document in `user-guide` then I can link to it from the `types` index page
